### ISSUE TITLE
Schema is always not up-to-date in 2.4.6 and MySQL 8.0.30 in some cases with real db fields

### ISSUE THREAD

**Issue by speller** [magento/magento2#37543]
### Preconditions and environment

- Magento version: 2.4.6
- MySQL version 8.0.30

### Steps to reproduce

1. Run `setup:upgrade`

### Expected result

The `setup:db:status` command reports that everything is up-to-date.

### Actual result

Running `setup:db:status` always reports thath the schema is not up-to-date.

```
bin/magento setup:db:status
Declarative Schema is not up to date
Run 'setup:upgrade' to update your DB schema and data.
```

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [X] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [X] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by speller** [comment]
The following workaround fixes the issue: https://github.com/magento/magento2/issues/36558#issuecomment-1334870008

**Comment by speller** [comment]
I double-checked and found that the issue is caused by the module `yotpo/module-yotpo-reviews`. By some reason, the `$elementStructuralData` value in the `\Magento\Framework\Setup\Declaration\Schema\Dto\ElementFactory::castGenericAttributes()` method doesn't have the `unsigned` array key on unsigned fields (i.e. `yotpo_orders_sync.yotpo_id`, `yotpo_product_sync.yotpo_id`, etc).
We're using version 4.0.33

**Comment by speller** [comment]
Probably, the `double unsigned` type is the issue.

**Comment by speller** [comment]
Yes, the issue is in the `\Magento\Framework\Setup\Declaration\Schema\Db\MySQL\Definition\Columns\Real::fromDefinition()` method that fails to process the `double unsigned` definition correctly:

<img width="926" alt="image" src="https://github.com/magento/magento2/assets/1829515/cefd23c6-4080-4a80-8b22-5273d0f32179">

The execution doesn't go into the if branch where the unsigned flag can be detected. So this is definitely a Magento bug.

**Comment by engcom-Bravo** [comment]
Hi @speller,

Thanks for your update.

Verified the issue on Magento 2.4-develop instance and Magento 2.4.6 still we are not able to reproduce the issue.Kindly refer the screenshots.


<img width="545" alt="Screenshot 2023-06-01 at 1 39 28 PM" src="https://github.com/magento/magento2/assets/51680745/0a2863c8-183c-4018-ab00-7d21cee94603">



<img width="557" alt="Screenshot 2023-06-01 at 1 39 35 PM" src="https://github.com/magento/magento2/assets/51680745/520109eb-cc7b-4323-aadd-cde323bfe611">


It looks like issue caused by `yotpo/module-yotpo-reviews`.and it is an third party module.

Kindly recheck the issue on Magento 2.4-develop instance and check if any 3rd party extensions / modules enabled is causing this issue.


Thanks,

**Comment by speller** [comment]
@engcom-Bravo 

Third-party modules can not do anything here except what is allowed by the framework. 

This is the definition of one of the columns that cause the issue:

```
        <column xsi:type="double" name="yotpo_id" unsigned="true" nullable="false" default="0" comment="Api Response Data - Yotpo ID"/>
```

Since Magento allows this and it causes problems, then this is a Magento issue. If the framework will not allow but the module will force by hacks then it will be the module issue. But the latter doesn't work here. The module uses the allowed schema and the framework can not handle this case properly.

**Comment by speller** [comment]
@engcom-Hotel 
What `setup:db:status` command says in your case?

**Comment by engcom-Hotel** [comment]
Hello @speller,

Below is the output of the `bin/magento setup:db:status` command:

![image](https://github.com/magento/magento2/assets/51681618/81533f61-7b5b-464e-bd1c-f149e5d57464)

And when we tried the above command after removing the `yotpo` modules then the output is as follows:

![image](https://github.com/magento/magento2/assets/51681618/8f5ae12c-23f8-4f21-83df-7ae62737f0d5)

Hence confirming the issue.

Thanks

**Comment by lbajsarowicz** [comment]
The setup:db:status mismatch reported here looks dependent on the specific DB version and column introspection (tested against MySQL 8.0.30). The linked PR reference (236) doesn't resolve to an actual magento2 PR, so that linkage looks incorrect. Can you provide your exact MySQL/MariaDB version and the declarative schema diff that triggers the mismatch?

**Comment by shuaiZend** [comment]
**Another distinct trigger of the same symptom: `utf8` / `utf8mb3` charset aliasing in `DbSchemaReader::getTableOptions()`**

@lbajsarowicz — exact versions and diff as requested.

### Environment

- Magento **2.3.7-p4** (reproduced and fixed here); code inspection of **2.4-develop** (checked 2026-09) shows the introspection side is still affected
- DB: Tencent **TXSQL 8.0.30** (MySQL 8.0.30 wire/behavior compatible) — applies to any vanilla **MySQL ≥ 8.0.28**
- Third-party modules whose `db_schema.xml` explicitly declares `charset="utf8"` / `collation="utf8_general_ci"` (in our case: Amasty_ShippingTableRates, Mageplaza_Smtp, Magenest_ZohocrmIntegration, Aheadworks_RewardPoints)

### Root cause

1. Since **MySQL 8.0.28**, `information_schema` reports the `utf8` charset as `utf8mb3` (`CHARACTER_SET_NAME`) and `utf8_general_ci` as `utf8mb3_general_ci` (`COLLATION_NAME`).
2. `DbSchemaReader::getTableOptions()` returns those raw values. Verified on 2.4-develop: there is **no `utf8mb3` normalization** anywhere on the introspection path (only the MariaDB ≥ 11.4 `FULL_COLLATION_NAME` handling exists), and `Dto\Table::getDiffSensitiveParams()` includes `charset` and `collation` (with zero `mb3` handling in the DTO either).
3. The comparator therefore sees declared `utf8` / `utf8_general_ci` vs introspected `utf8mb3` / `utf8mb3_general_ci` → **every `utf8` table is flagged as modified on every `setup:upgrade` run** → repeated `ALTER TABLE ... CONVERT TO CHARACTER SET` suggestions, and when FKs are involved the re-creation fails with `SQLSTATE[HY000]: General error: 1826 Duplicate foreign key constraint name`.

This is a different trigger than the `yotpo/module-yotpo-reviews` `double unsigned` / `Real::fromDefinition()` case discussed earlier in this thread.

### Proposed fix

`utf8` and `utf8mb3` are aliases of the same charset, so the introspection side should canonicalize back to the form used in declarations:

```diff
--- a/lib/internal/Magento/Framework/Setup/Declaration/Schema/Db/MySQL/DbSchemaReader.php
+++ b/lib/internal/Magento/Framework/Setup/Declaration/Schema/Db/MySQL/DbSchemaReader.php
@@ -84,7 +84,17 @@
             ->where('TABLE_SCHEMA = ?', $dbName)
             ->where('TABLE_NAME = ?', $tableName);

-        return $adapter->fetchRow($stmt);
+        $row = $adapter->fetchRow($stmt);
+
+        // MySQL >= 8.0.28 reports the utf8 charset as utf8mb3 in information_schema.
+        // Normalize it back to the canonical form used in db_schema.xml declarations
+        // (utf8 and utf8mb3 are aliases of the same charset), otherwise the table-level
+        // diff (charset/collation are part of getDiffSensitiveParams()) flags every
+        // utf8 table as modified on every setup:upgrade run.
+        if (isset($row['charset'])) {
+            $row['charset'] = str_replace('utf8mb3', 'utf8', $row['charset']);
+        }
+        if (isset($row['collation'])) {
+            $row['collation'] = str_replace('utf8mb3_', 'utf8_', $row['collation']);
+        }
+
+        return $row;
     }
```

### Verification (production-shaped dataset, ~420 tables incl. 4 third-party declarative-schema modules)

- **Without the patch**: every `setup:upgrade` run emits the same set of ALTERs; FK re-creation aborts with `ERROR 1826 Duplicate foreign key constraint name`; `bin/magento db:status` never reports up-to-date.
- **With the patch**: 3 consecutive `setup:upgrade` runs are idempotent (0 schema ALTERs) and `bin/magento db:status` reports `All modules are up to date.`

### Relation to the recent AC-16562 changes (`DefaultCharsetCollationMap` on 2.4-develop)

The new version-aware defaults (declared side switches to `utf8mb4` on MySQL ≥ 8.0.29) do **not** remove this problem:

- tables explicitly declaring `charset="utf8"` in `db_schema.xml` still introspect as `utf8mb3` → false diff remains on 2.4-develop;
- legacy tables created with `utf8` and *without* explicit charset would now compare declared `utf8mb4` vs introspected `utf8mb3` — that direction (implicit charset migration via schema diff) should arguably be surfaced explicitly rather than silently ALTERed, but in either case the introspection side needs alias-aware canonicalization for consistent comparison.
